rholang/reference_doc/data_structures/README.md

```diff
@@ -1,73 +1,117 @@
+Rholang supports the four basic data structures: `List`, `Map`, `Set`, `Tuple`, i.e. the type
+```rholang
+  DataStruct ::= List | Map | Set | Tuple
 ```
 
-Examples:
+### List
+A Rholang *list* is an ordered collection of 0 or more processes. The syntax is:
+```rholang
+  List ::= "[" "]" | "[" Procs "]"
 ```
-TODO include the examples from ../../examples/tut-tuples-methods.rho
+where a collection of processes `Procs` is given by
+```rholang
+  Procs ::= Proc | Proc "," Procs
 ```
 
-### List
+- more info
 
 #### List Concatenation
 While not technically a method, lists support concatenation using the `++` binary operator. Both operands must be lists.
 
 Example:
 ```rholang
-[1, 3, 5] ++ [2, 4, 6] -> [1, 3, 5, 2, 4, 6]
+  [1, 3, 5] ++ [2, 4, 6] => [1, 3, 5, 2, 4, 6]
+```
+
+### Map
+A Rholang *map* is an unordered collection of 0 or more key-value pairs; both keys and values are processes. The syntax is:
+```rholang
+      Map ::= "{" "}" | "{" KVPairs "}"
+   KVPair ::= Proc ":" Proc
+  KVPairs ::= KVPair | KVPair "," KVPairs
 ```
 
+The presentation of a map in Rholang is normalized by its keys. E.g. the map `{ 2:5 , 0:2 , 1:1 }` is presented as `{ 0:2 , 1:1 , 2:5 }`.
+
 ### Set
+A Rholang *set* is an unordered collection of 0 or more processes. The syntax is:
+```rholang
+  Set ::= "Set(" ")" | "Set(" Procs ")"
+```
 
-### Map
+The presentation of a set in Rholang is normalized
 
+### Tuple
+A Rholang *tuple* is an ordered collection of 1 or more processes. The syntax is:
+```rholang
+ Tuple ::= "(" Proc ",)" | "(" Proc "," Procs ")"
+```
 
+#### Examples of the syntax
+- `[ ]` empty list
+- `[ Set(true) , @19!(23) , { 1:2 , "a":2 }.get("a") ]` a length 3 list
+- `{ }` empty map
+- `{ 0:x , 1:y!(0) , "abc":"a" , Nil:{for(y <- x){*y}} }` a size 4 map
+- `Set( )` empty set
+- `Set( for( _ <<- @0 ){ Nil } , "a" , 42 , [ *chan , 1 ] , true )` a size 5 set
+- `(true , 1 , chan!(2) , [ 1 , Nil ] , "a" , { ( Nil ,):0 } )` a 6-tuple
+- `(Nil ,)` a 1-tuple (notice the trailing comma)
 
+Examples:
+```
+TODO include the examples from ../../examples/tut-tuples-methods.rho
+```
 
 ## Methods
+All method calls have the same syntactic form:
+```rholang
 
-Each method needs its own technical description.
-
-- `nth`
-- `toByteArray`
-- `hexToBytes`
-- `union`
-- `diff`
-- `add`
-- `delete`
-- `contains`
-- `get`
-- `getOrElse`
-- `set`
-- `keys`
-- `size`
-- `length`
-- `slice`
-
-(Not yet in the interpreter)
-- `confine`
-
-
-Some standard form for these: function, inputs, outputs. Easy to read, uniform.
+  MethodCall ::= Data "." MethodName "(" Args ")"
+        Args ::= "" | Procs
+```
+where `MethodName` is a valid method name token and `Args` is a valid (for the method) list of process arguments.
+
+Some methods may be called without argument(s), while others require an acknowledgement channel or other parameters.
+
+##### `confine`
+- `confine`: what is confined? (Not yet in the interpreter)
+
+##### `toByteArray()`
```

Comment:
Could you add `toUtf8Bytes` here as well`? Here's the PR re this method https://github.com/rchain/rchain/pull/1609, you can just copy paste definition.